src/types.rs

```diff
@@ -258,6 +259,149 @@ impl fmt::Display for ResolvedType {
     }
 }
 
+/// Simfony type with type aliases.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub struct AliasedType(AliasedInner);
+
+/// Type alias or primitive.
+///
+/// Private struct to allow future changes.
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+enum AliasedInner {
+    /// Type alias
+    Alias(Identifier),
+    /// Type primitive
+    Inner(TypeInner<Arc<AliasedType>>),
+}
+
+impl AliasedType {
+    /// Create a type alias from the given `identifier`.
+    pub fn alias(identifier: Identifier) -> Self {
+        Self(AliasedInner::Alias(identifier))
+    }
+
+    /// Resolve all aliases in the type based on the given map of `aliases` to types.
+    pub fn resolve<F>(&self, mut get_alias: F) -> Result<ResolvedType, Identifier>
+    where
+        F: FnMut(&Identifier) -> Option<ResolvedType>,
+        // The borrow checker doesn't let me return Option<ResolvedType>
+        // F: FnMut(&Identifier) -> Option<&ResolvedType>,
+        // note: due to current limitations in the borrow checker, this implies a `'static` lifetime
```

Comment:
In 43db8c9aa2b4d85d70706d1eee97e18a9618cc53:

FYI if you add a `'s` lifetime parameter to `resolve`, change `&self` to `&'s self`, then change `&Identifier` to `&'s Identifier` and `Option<ResolvedType>` to `Option<&'s ResolvedType>`, things work.

But this doesn't really gain you anything. It lets you drop the `.cloned()` on the closure you pass to `resolve` but then you need to add a `.cloned()` within `resolve`. There's no way to return a reference from `resolve` itself since sometimes you construct new `ResolvedType`s and return it.

I would suggest just dropping this comment because it invites people to try to mess around with it, and the problem isn't actually with the borrowchecker.